Add suggestion to remove if in let..else block

edward-shen opened the pull request:
Adds an additional hint to failures where we encounter an else keyword while we're parsing an if-let expression.

This is likely that the user has accidentally mixed if-let and let..else together.

Fixes #103791.

edward-shen commented on tests/ui/rfc-2497-if-let-chains/ensure-that-let-else-does-not-interact-with-let-chains.stderr:
```diff
@@ -37,6 +37,11 @@ help: add a block here
    |
 LL |     if let Some(n) = opt else {
    |                         ^
+help: There may be an extra `if` before the `let...else`
+  --> $DIR/ensure-that-let-else-does-not-interact-with-let-chains.rs:24:5
+   |
+LL |     if let Some(n) = opt else {
+   |     ^^
```
Related test source code: [rfc-2497-if-let-chains/ensure-that-let-else-does-not-interact-with-let-chains.rs](https://github.com/rust-lang/rust/blob/cf2dff2/tests/ui/rfc-2497-if-let-chains/ensure-that-let-else-does-not-interact-with-let-chains.rs).

This behavior is correct as this hint should only appear when it is ambiguous if the user is try to use an `if-let` or an `let..else`.

The other two cases that this test handles are:
```rust
    if let Some(n) = opt && n == 1 else {
    //~^ ERROR this `if` expression is missing a block after the condition
        return;
    };
    if let Some(n) = opt && let another = n else {
    //~^ ERROR this `if` expression is missing a block after the condition
        return;
    };
```

which are clear attempts at `if-let` chains.

This also indicates that we don't emit this help on `while-let` -- this wasn't in scope of the original issue so I took a conservative approach in this PR.

compiler-errors commented:
I can take a look at this tomorrow.

r? @compiler-errors

est31 commented on tests/ui/let-else/accidental-if.stderr:
```diff
@@ -0,0 +1,19 @@
+   |
+LL |     if let Some(y) = x else {
+   |                       ^
+help: There may be an extra `if` before the `let...else`
```
This can be misinterpreted in the way that the statement is saying that you are allowed to put `if` in front of `let...else`, instead of what the statement *intends* to say, of the author potentially having mistakenly put an `if` in front of `let...` else.

I wonder if there are better ways to write this suggestion that can't be misinterpreted.

edward-shen replied:
Thanks for the feedback! How about:

```
help: Remove the `if` to construct a `let...else` statement
```

I think that's a little more direct while still sounding neutral.

est31 replied:
@edward-shen that is perfect!

Rageking8 commented on compiler/rustc_error_messages/locales/en-US/parse.ftl:
```diff
@@ -238,6 +238,7 @@ parse_const_let_mutually_exclusive = `const` and `let` are mutually exclusive
 
 parse_invalid_expression_in_let_else = a `{$operator}` expression cannot be directly assigned in `let...else`
 parse_invalid_curly_in_let_else = right curly brace `{"}"}` before `else` in a `let...else` statement not allowed
+parse_extra_if_in_let_else = Remove the `if` to construct a `let...else` statement
```
```suggestion
parse_extra_if_in_let_else = remove the `if` to construct a `let...else` statement
```
Sorry for being this pedantic, but AFAIK the rest of the help messages begin with a lowercase. So maybe uniformity can be preserved here. Thanks.

compiler-errors commented on compiler/rustc_parse/src/errors.rs:
```diff
@@ -337,7 +337,7 @@ pub(crate) struct IfExpressionMissingThenBlock {
     pub if_span: Span,
     #[subdiagnostic]
-    pub sub: IfExpressionMissingThenBlockSub,
+    pub sub: Vec<IfExpressionMissingThenBlockSub>,
```
Instead of having a vector of subdiagnostics, could you make it something like:

```
pub sub: IfExpressionMissingThenBlockSub,
pub let_else_sub: Option<ExtraIfInLetElseSub>,
```

And then pass either `None` or `Some(ExtraIfInLetElseSub(span))` into the `IfExpressionMissingThenBlock` diagnostic?

That makes the code a bit easier to understand IMO.

compiler-errors commented on compiler/rustc_error_messages/locales/en-US/parse.ftl:
```diff
@@ -238,6 +238,7 @@ parse_const_let_mutually_exclusive = `const` and `let` are mutually exclusive
 
 parse_invalid_expression_in_let_else = a `{$operator}` expression cannot be directly assigned in `let...else`
 parse_invalid_curly_in_let_else = right curly brace `{"}"}` before `else` in a `let...else` statement not allowed
+parse_extra_if_in_let_else = remove the `if` to construct a `let...else` statement
```
```suggestion
parse_extra_if_in_let_else = remove the `if` if you meant to write a `let ... else` statement
```

compiler-errors replied:
Maybe this wording is a bit better? Idk

edward-shen replied:
Re: the space between let and else, the line above it already doesn't include them. Not sure if you want me to stay consistent?

edward-shen replied:
Re: wording, I searched around, looks like that's more consistent with what else is in here, thanks.

compiler-errors replied:
uhhh i guess just keep it as-is, just apply the wording changes if you think it's clearer

edward-shen replied:
Yeah, I keep forgetting to rebless the tests. There'll be one more force-push

compiler-errors reviewed (commented):
This looks pretty good to me as is. Please address nits (or tell me why I'm wrong :laughing:) and I think this can get r+'d soon.

compiler-errors commented on compiler/rustc_error_messages/locales/en-US/parse.ftl:
```diff
@@ -238,6 +238,7 @@ parse_const_let_mutually_exclusive = `const` and `let` are mutually exclusive
 
 parse_invalid_expression_in_let_else = a `{$operator}` expression cannot be directly assigned in `let...else`
 parse_invalid_curly_in_let_else = right curly brace `{"}"}` before `else` in a `let...else` statement not allowed
+parse_extra_if_in_let_else = remove the `if` if you meant to write `let...else` statement
```
is this missing an "a" before let...else?

edward-shen replied:
Yes.

edward-shen committed:
Add suggestion to remove if in let...else block

Adds an additional hint to failures where we encounter an else keyword
while we're parsing an if-let block.

This is likely that the user has accidentally mixed if-let and let...else
together.

compiler-errors commented:
@bors r+ rollup

bors commented:
:pushpin: Commit a8b77cfe5464e29b25389593ced9d080bf0dd6c8 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors a8b77cfe5464e29b25389593ced9d080bf0dd6c8 -->
<!-- homu: {"type":"Approved","sha":"a8b77cfe5464e29b25389593ced9d080bf0dd6c8","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

estebank commented on compiler/rustc_parse/src/errors.rs:
```diff
@@ -348,6 +350,13 @@ pub(crate) enum IfExpressionMissingThenBlockSub {
     AddThenBlock(#[primary_span] Span),
 }
 
+#[derive(Subdiagnostic)]
+#[help(parse_extra_if_in_let_else)]
+pub(crate) struct IfExpressionLetSomeSub {
+    #[primary_span]
+    pub if_span: Span,
+}
```
This should have been a structured suggestion, not a help, so that rustfix can apply the change. The missing block help isn't a structured suggestion because we didn't have any code for it (but arguably it should have been as well).

edward-shen replied:
Is this just changing the `help` to a `suggestion`? If so I can address this in a follow-up PR.

compiler-errors replied:
Yup

compiler-errors commented on compiler/rustc_parse/src/parser/expr.rs:
```diff
@@ -11,15 +11,15 @@ use crate::errors::{
+    IfExpressionMissingThenBlockSub, InvalidBlockMacroSegment, InvalidComparisonOperator,
+    InvalidComparisonOperatorSub, InvalidInterpolatedExpression, InvalidLiteralSuffixOnTupleIndex,
+    InvalidLogicalOperator, InvalidLogicalOperatorSub, LabeledLoopInBreak, LeadingPlusNotSupported,
+    LeftArrowOperator, LifetimeInBorrowExpression, MacroInvocationWithQualifiedPath,
```
Alternatively, let's replace this with a glob import

est31 replied:
Note: this is nothing that this particular PR has caused. The pattern of having a giant use at the top just creates this noise in PRs like this.

est31 replied:
Let's discuss it [on zulip](https://rust-lang.zulipchat.com/#narrow/stream/147480-t-compiler.2Fwg-diagnostics).
